### PR TITLE
Auto Update Framework

### PR DESCRIPTION
**Pull request by mbhavi**
This PR includes changes for Auto Update of Brackets.
The changes are with respect to:
1) Windows implementation
2) Mac implementation

ping @swmitra @nethip for review

### CODE REVIEW

File: appshell/appshell_extensions.cpp [hunk 1]
```diff
@@ -507,7 +508,35 @@ class ProcessMessageDelegate : public ClientHandler::ProcessMessageDelegate {
                 error = CopyFile(src, dest);
                 // No additional response args for this function
             }
-        } else if (message_name == "GetDroppedFiles") {
+        } else if (message_name == "SetUpdateParamsAndRunUpdate") {
+			// Parameters:
```
Comment (nethip) [inline appshell/appshell_extensions.cpp]: nit: Please fix the indentation.

File: appshell/appshell_extensions.cpp [hunk 2]
```diff
@@ -507,7 +508,35 @@ class ProcessMessageDelegate : public ClientHandler::ProcessMessageDelegate {
                 error = CopyFile(src, dest);
                 // No additional response args for this function
             }
-        } else if (message_name == "GetDroppedFiles") {
+        } else if (message_name == "SetUpdateParamsAndRunUpdate") {
```
Comment (nethip) [inline appshell/appshell_extensions.cpp]: Where are you calling this method from in your extension?

File: appshell/update_mac.mm [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        //Util paths
+        NSString* hdiPath = @"/usr/bin/hdiutil";
+        NSString* shPath = @"/bin/sh";
+        NSString* nohupPath = @"/usr/bin/nohup";
+    
+        // Unmount the already existing disk image, if any
+        NSArray* pArgs = [NSArray arrayWithObjects:@"detach", mountPoint,  @"-force", nil];
+        
+        int retval = RunScript(hdiPath, pArgs, true);
+        while (retval != 1) {
```
Comment (nethip) [inline appshell/update_mac.mm]: When do you exit from the loop?

File: appshell/appshell_extensions.cpp (hunk 2 above)
Comment (nethip) [reply appshell/appshell_extensions.cpp]: @mbhavi Can you help me point to the exact location in your PR https://github.com/adobe/brackets/pull/14177. You could click on the line where this is getting called and copy paste the URL from the browser, here.

Comment (nethip) [reply appshell/appshell_extensions.cpp]: Thanks @mbhavi Got it!

File: appshell/appshell_extensions.cpp (hunk 1 above)
Comment (mbhavi) [reply appshell/appshell_extensions.cpp]: Addressed in ad0d490

**Comment by mbhavi** [comment]
@nethip , I have addressed and committed all the required changes. Also, I have replied to all the comments.
The PR is ready for review again.

**Comment by lokwani** [comment]
Installer LGTM

File: Gruntfile.js [hunk 4]
```diff
@@ -192,6 +192,19 @@ module.exports = function (grunt) {
                     }
                 ]
             },
+            "winInstallerDLLs": {
+                "files": [
+                    {
+                        "flatten"   : true,
+                        "expand"    : true,
+                        "src"       : [
+                            "installer/win/LaunchBrackets/LaunchBrackets/bin/Release/LaunchBrackets.CA.dll",
```
Comment (nethip) [inline Gruntfile.js]: Where is the source for these DLLs?

File: installer/win/Brackets.wxs [hunk 5]
```diff
@@ -114,18 +114,70 @@
...
+                            Key="directory\shell\!(loc.ProductName)"
+            />
+        </Property>
+
+        <SetProperty Id="CURRENTPATH"
+             Before="CostInitialize" 
+             Sequence="execute"
+             Value="[%PATH]" />
+
+        <Binary Id="InstallLocationChecker.dll" SourceFile="BracketsConfigurator.CA.dll" />
```
Comment (nethip) [inline installer/win/Brackets.wxs]: The source for this and how these DLLs are built?

Comment (nethip) [reply installer/win/Brackets.wxs]: Found them below. Please ignore the comment.

File: installer/win/BracketsConfigurator/BracketsConfigurator/BracketsConfigurator.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
+using System;
```
Comment (nethip) [inline installer/win/BracketsConfigurator/BracketsConfigurator/BracketsConfigurator.cs]: Copyright missing?

File: installer/win/LaunchBrackets/LaunchBrackets/LaunchBrackets.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
+using System;
```
Comment (nethip) [inline installer/win/LaunchBrackets/LaunchBrackets/LaunchBrackets.cs]: Copyright missing?

File: Gruntfile.js (hunk 4 above)
Comment (nethip) [reply Gruntfile.js]: Ignore this comment.

File: installer/win/BracketsConfigurator/BracketsConfigurator/BracketsConfigurator.cs (hunk 6 above)
Comment (mbhavi) [reply installer/win/BracketsConfigurator/BracketsConfigurator/BracketsConfigurator.cs]: @gavankar , please address this

File: installer/win/LaunchBrackets/LaunchBrackets/LaunchBrackets.cs (hunk 7 above)
Comment (mbhavi) [reply installer/win/LaunchBrackets/LaunchBrackets/LaunchBrackets.cs]: @gavankar , please address this.

Comment (gavankar) [reply installer/win/LaunchBrackets/LaunchBrackets/LaunchBrackets.cs]: @nethip we have never had a copyright in the installer files so far(installer/win/Brackets.wxs), should we be adding one now?

File: installer/win/BracketsConfigurator/BracketsConfigurator/BracketsConfigurator.cs (hunk 6 above)
Comment (gavankar) [reply installer/win/BracketsConfigurator/BracketsConfigurator/BracketsConfigurator.cs]: @nethip we have never had a copyright in the installer files so far(installer/win/Brackets.wxs), should we be adding one now?

Comment (nethip) [reply installer/win/BracketsConfigurator/BracketsConfigurator/BracketsConfigurator.cs]: Yes please go ahead.

Comment (gavankar) [reply installer/win/BracketsConfigurator/BracketsConfigurator/BracketsConfigurator.cs]: @mbhavi, Please accept the following pull request which has this fix -
https://github.com/mbhavi/brackets-shell/pull/2

**Comment by mbhavi** [comment]
@nethip , addressed review comments and making auto update no-op on Linux, in 3f6d92d

File: installer/win/BracketsConfigurator/BracketsConfigurator/BracketsConfigurator.cs (hunk 6 above)
Comment (mbhavi) [reply installer/win/BracketsConfigurator/BracketsConfigurator/BracketsConfigurator.cs]: @gavankar , I have merged your PR. @nethip , please find the review comments addressed here, as part of this merge.

**Comment by nethip** [comment]
@mbhavi @gavankar Are the custom action DLLs like `BracketsExecutor` part of this PR? Can you create a `wiki` post merging this PR as to what the steps to generate the DLLs.

Also @mbhavi post merging this PR, can you raise a PR with unit tests for the auto update process?

**Comment by gavankar** [comment]
@nethip, the steps to create BracketsConfigurator.CA.dll and LaunchBrackets.CA.dll have been added to the "build-win" grunt task, these steps are completely automated through grunt. Do we still require a wiki about the steps for generating the dlls?

**Comment by nethip** [comment]
@gavankar Thanks for pointing out! Pretty neat! No need of adding a wiki for this.

This PR looks good. LGTM

@mbhavi Thanks for your efforts in making this happen and being patient in addressing the review comments :smile: . Really appreciate it. And great job :100:  Please address the review comments in Brackets as well. Thanks!

I would merge this after @ayushrathi15 confirms the results from sanity testing.

**Pull request merged**